### PR TITLE
`collect-outputs` util

### PR DESCRIPTION
**Pull request by jprokos26**
Can be utilized to generate evaluation tables automatically utilizing various flags.
Organize by performer, attack type, or any other config field.
Additionally clean up your output directory automatically.
- Remove folders with just `armory-log.txt` and `colored-log.txt`
- Remove pairs of duplicate png images from `saved_samples`
- Removed files are placed in `~/.armory/outputs/.cleaned`

Example:
After running:
```shell
cd ~/git/twosixlabs/gard-evaluatsion/MITM_sleeper_agent/jprokos26
python3 -m armory utils collect-outputs --collate --glob "*Poison*.json" --sort --filter "diffusion"
cp -Lr ~/.armory/results/SleeperAgentAttack* .
git commit -am "Adding diffusion sleeper results"
```
_Note, cwd affects href link for config paths_
See https://github.com/twosixlabs/gard-evaluations/tree/eval7-jp-MITM_sleeper/MITM_sleeper_agent/jprokos26 for output

### CODE REVIEW

File: armory/cli/tools/collect_outputs.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+    """return a row of the table for a CARLA adversarial patch attack"""
+    row = [None] * len(CARLA_HEADERS)
+    for i, parse in enumerate(CARLA_HEADERS.values()):
+        try:
+            value = parse(json_data)
+        except Exception as e:
+            log.error(
+                f"Error parsing {list(CARLA_HEADERS.keys())[i]} from {filepath}:\n{e}"
+            )
+            breakpoint()
```
Comment (jprokos26) [inline armory/cli/tools/collect_outputs.py:187]: remove before merge

File: armory/cli/tools/collect_outputs.py [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+        return None
+    if isinstance(rate, list):
+        return _round_rate(sum(rate) / len(rate), places)
+    return round(rate, places)
+
+
+def _unknown_mean_key_helper(d, key: str, places=2) -> str:
+    # Don't know which key it will be under, so try both
+    if key == "carla_od_disappearance_rate":
+        breakpoint()
```
Comment (jprokos26) [inline armory/cli/tools/collect_outputs.py:140]: remove before merge

File: armory/cli/tools/collect_outputs.py [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+        target = clean_dir / d.relative_to(output_dir)
+        target.parent.mkdir(parents=True, exist_ok=True)
+        d.rename(target)
+    # delete all pairs of .png files NAME.png and NAME_adv.png where both files are equivalent
+    # (i.e. NAME.png is a copy of NAME_adv.png)
+    log.info("Cleaning up duplicate .png files")
+
+    pngs = list(
+        filter(
+            lambda p: "MITM" in str(p),
```
Comment (jprokos26) [inline armory/cli/tools/collect_outputs.py:44]: Fix

File: armory/cli/tools/collect_outputs.py [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+    )
+    unlinks = set()
+    for png in simple_progress_bar(pngs, msg="Checking equality..."):
+        adv = png.parent / (png.stem + "_adv.png")
+        if adv.exists() and Image.open(png).tobytes() == Image.open(adv).tobytes():
+            # log.debug(f"Deleting {png}")
+            unlinks.add(png)
+            unlinks.add(adv)
+        # else:
+        # log.debug(f"Keeping {adv}")
```
Comment (jprokos26) [inline armory/cli/tools/collect_outputs.py:56]: clean up

File: armory/cli/tools/collect_outputs.py [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+    if isinstance(headers[0], list):
+        return "\n\n".join([_parse_markdown_table(h, r) for h, r in zip(headers, rows)])
+    table = "\n".join(
+        [
+            "|".join(headers),
+            "|".join(["---"] * len(headers)),
+            *["|".join(map(str, row)) for row in rows],
+        ]
+    )
+    return table
```
Comment (jprokos26) [inline armory/cli/tools/collect_outputs.py:311]: TODO pad with appropriate padding so output is more readable in text editor

File: docs/utils.md [hunk 6]
```diff
@@ -80,3 +81,52 @@ See [our write-up](https://docs.google.com/document/d/1_8_nRQmHhK5ieHNcGhoRtBZcR
...
+
+
+```shell
+Convert runs from the output directory into tables
+
+options:
+  -h, --help            show this help message and exit
+  --glob GLOB, -g GLOB  Glob pattern to match json outputs. Defaults to `*.json`.
+  --output OUTPUT, -o OUTPUT
+                        Path to output tables. Defaults to /home/jonathan.prokos/.armory/results/ATTACK.md where str format placeholder is replaced with ATTACK name if supplied.
```
Comment (jprokos26) [inline docs/utils.md:105]:
```suggestion
                        Path to output tables. Defaults to ARMORY_OUTPUT_DIR/../results/ATTACK.md where str format placeholder is replaced with ATTACK name if supplied.
```

File: armory/cli/tools/collect_outputs.py [hunk 7]
```diff
@@ -0,0 +1,581 @@
...
+    ignore_keys = {
+        "optimizer",
+        "targeted",
+        "verbose",
+        "patch_base_image",
+        "patch",
+        "device_name",
+        "batch_size",
+    }
+    _ = d["config"]["attack"]["kwargs"].pop("patch_mask", None)  # remove patch_mask
```
Comment (christopherwoodall) [inline armory/cli/tools/collect_outputs.py:114]: Can `patch_mask` be added to `ignore_keys` instead of popping?

**Review by christopherwoodall — APPROVED** [review]
The suggestion below does not prohibit the inclusion of this branch.

File: armory/cli/tools/collect_outputs.py (hunk 7 above)
Comment (jprokos26) [reply armory/cli/tools/collect_outputs.py:114]: The kwargs section needs cleaned up; `patch_mask` should be included in the output. Will address shortly

File: armory/cli/tools/collect_outputs.py [hunk 8]
```diff
@@ -0,0 +1,581 @@
...
+            else:
+                collation[str(tgt_collate)].add(attack_name)
+        tables[attack_name].append(row)
+    log.info(f"Parsed {len(tables)} results.")
+
+    # collate results
+    if args.collate is not False:
+        for tgt_collate, attacks in collation.items():
+            # make sure all attacks have the same headers
+            first = attacks.pop()
```
Comment (jprokos26) [inline armory/cli/tools/collect_outputs.py:543]: Pop removes the element, need to add it back!

Comment (jprokos26) [reply armory/cli/tools/collect_outputs.py:543]: Addressed by https://github.com/twosixlabs/armory/pull/1986/commits/48f1efe5f74b2b266a450c2fc1a2be4c2b87705a

File: armory/cli/tools/collect_outputs.py [hunk 9]
```diff
@@ -0,0 +1,582 @@
...
+    while filepath is None and len(fallthrough) > 0:
+        filepath = _get_dict_value(fallthrough.pop(0), d, default_value=None)
+    if filepath is None:
+        # can't find config filepath, just return result filepath as %s
+        return "%s"
+    p_filepath = Path(filepath)
+    if p_filepath.is_absolute():
+        common = os.path.commonpath([Path.cwd(), p_filepath])
+        filepath = p_filepath.relative_to(common)
+    return f"[{filepath}]({filepath})"
```
Comment (jprokos26) [inline armory/cli/tools/collect_outputs.py:86]: Filepath can be `-` if config came from stdin; fixup what gets return in this case

**Comment by jprokos26** [comment]
Addressed by https://github.com/twosixlabs/armory/pull/1986/commits/723c810c601b72f36d4db2bc5894b8f53692dcc0.

Stale:
Add `runtime` field. Can be extracted from `armory-log.txt`
```bash
for file in SleeperAgentAttack/*/armory-log.txt; do echo -n $(basename $(find $(dirname $file) -name "*.json")); echo -n " - "; tail -n 1 $file | cut -d" " -f3; done
```

**Comment by jprokos26** [comment]
If defense is `null`, should grab other kwargs
https://github.com/twosixlabs/armory/blob/master/scenario_configs/eval6/poisoning/sleeper_agent/baseline_defenses/cifar10_sleeper_agent_p10_perfect_filter.json#L58-L64

File: armory/cli/tools/collect_outputs.py (hunk 9 above)
Comment (jprokos26) [reply armory/cli/tools/collect_outputs.py:86]: Addressed by https://github.com/twosixlabs/armory/pull/1986/commits/723c810c601b72f36d4db2bc5894b8f53692dcc0

File: armory/cli/tools/collect_outputs.py (hunk 7 above)
Comment (jprokos26) [reply armory/cli/tools/collect_outputs.py]: Addressed by https://github.com/twosixlabs/armory/pull/1986/commits/723c810c601b72f36d4db2bc5894b8f53692dcc0

**Review by swsuggs — APPROVED** [review]
Acknowledging that this tool is not comprehensive but will still be useful for eval 8.  The JATIC armory-library will accomplish similar goals via MLFlow.

**Pull request merged**